### DISCUSSION TITLE
how to add keyword without quotes

### DISCUSSION THREAD

**Discussion by Yokozuna59** [eclipse-langium/langium#1081] [Q&A] [answered]
If I have the following langium grammar:

```langium
grammar Example

entry Example:
    "example" name=ID
;

hidden terminal WS: /\s+/;
terminal ID: /[\s\S]+/;
```

Using `langium-cli`, it'll generate `example` as a keyword and would be highlighted differently and suggested when typing. But by setting whitespaces as hidden rules, then it would be fine to combine `example` with `ID`.

```
examplewow
```

That's why I thought of this:

```langium
grammar Example

entry Example:
    EXAMPLE name=ID
;

hidden terminal WS: /\s+/;
terminal EXAMPLE: /example(?!\S)/;
terminal ID: /[\s\S]+/;
```

Which restricts that there shouldn't be a non-whitespace letter after `example`, but now it's not defined as a keyword.

Is there a way to redefine it as keyword? Or another way to force `example` to not be combined with `ID`? Or a better approach?

**Comment by msujew** [comment, marked as answer]
I'm unsure what to recommend here. Mostly due to the question: What constitutes a keyword if not a well-known word that is separated from other words using a non-word character? I.e. either whitespace or other symbols such als brackets, braces or parenthesis.

So, in theory, you can just declare the keyword as normal and override the pattern in the token builder. That would probably accomplish what you're looking for the best.

You could also keep the regex terminal and then add the corresponding parts to the completion provider. Note that there's very little difference between keywords and other terminals during the language server runtime. It's basically only the completion that is provided for keywords. Non cross reference terminals don't get completion support.

**Reply by Yokozuna59** [reply]
> So, in theory, you can just declare the keyword as normal and override the pattern in the token builder. That would probably accomplish what you're looking for the best.

I guess I'll take this approach because:

- It doesn't need much work since I'm familiar with it.
- A bit clearer in grammar.
- Explicitly show that the work is keyword.

But it appears that if it's not a rule, then it won't go to token builder so that I can override the pattern.

**Reply by Yokozuna59** [reply]
I'm not really sure my question was answered :)

edit: when will the new version be released? I'm thrilled to try out railroad with other features :)

**Reply by spoenemann** [reply]
Hi, we're planning to release v2.0.0 by end of July.

**Reply by Yokozuna59** [reply]
> But it appears that if it's not a rule, then it won't go to token builder so that I can override the pattern.

Sorry, I overridden the `buildTerminalToken` by mistake.
